src/lib/autoinstall/script.rb

```diff
@@ -0,0 +1,368 @@
+    end
+  end
+
+  class ScriptFeedback
+    # feedback settings
+    # @return [:message | :warning | :error | :popup | :no]
+    attr_reader :value
+
+    def initialize(hash)
+      @value = if hash["feedback"]
+        case hash["feedback_type"]
+        when "message" then :message
+        when "warning" then :warning
+        when "error" then :error
+        when "", nil then :popup
+        else
+          raise "Unknown feedback type value #{hash["feedback_type"].inspect}"
+        end
+      else
+        :no
+      end
+    end
+
+    def to_hash
+      case @value
+      when :message then { "feedback" => true, "feedback_type" => "message" }
+      when :warning then { "feedback" => true, "feedback_type" => "warning" }
+      when :error then { "feedback" => true, "feedback_type" => "error" }
+      when :popup then { "feedback" => true, "feedback_type" => "" }
+      when :no then { "feedback" => false, "feedback_type" => "" }
+      else
+        raise "Unknown value '#{value.inspect}"
+      end
+    end
+  end
+
+  # Scripts that are executed by YaST ( so all expect init scripts )
+  class ExecutedScript < Script
+    # Sets if feedback after script finish should be shown
+    # @return [ScriptFeedback]
+    attr_reader :feedback
+
+    # Interpreter to use to run script.
+    # @return [String]
+    attr_reader :interpreter
+
+    # Notification message during script run
+    # @return [String] empty string if no notification should be shown
+    attr_reader :notification
+
+    # Params passed to script
+    # @return [Array]
+    attr_reader :params
+
+    # Reruns script even if already ran
+    # @return [Boolean]
+    attr_reader :rerun
+
+    def to_hash
+      res = super
+      res["interpreter"] = interpreter
+      res["notification"] = notification
+      res["param-list"] = params
+      res["rerun"] = rerun
+      res.merge(feedback.to_hash)
+    end
+
+    def initialize(hash)
+      super
+
+      @feedback = ScriptFeedback.new(hash)
+      @interpreter = hash["interpreter"] || "shell"
+      @notification = hash["notification"] || ""
+      @params = hash["param-list"] || []
+      @rerun = !!hash["rerun"]
+    end
+
+    DEBUG_FLAG_MAP = {
+      "shell" => "-x",
+      "perl" => "-w",
+      "ruby" => "-w"
+    }
+    INTERPRETER_MAP = {
+      "shell" => "/bin/sh", # TODO: why not bash? at least user can now specify interpreter he wants
+      "perl" => "/usr/bin/perl",
+      "python" => "/usr/bin/python"
+    }
+    # Runs the script
+    # @return [Boolean] if exit code is zero
+    def execute
+      return if already_run? && !rerun
+
+      # TODO: maybe own class for interpreters?
+      cmd = INTERPRETER_MAP[interpreter] || interpreter
+      debug_flag = debug ? (DEBUG_FLAG_MAP[interpreter] || "") : ""
+      params_s = params.join(" ") # shell escaping is up to user, see documentation
+
+      bash_path = Yast::Path.new(".target.bash")
+      res = Yast::SCR.Execute(bash_path,
```

Comment:
Would `Yast::Execute` (Cheetah) offer any benefit?

Reply:
sadly not in this case as it intentionally force user to do shell escaping, so we need to execute cheetah as "sh -c '...'" so many cheetah security features are disabled and same for exceptions.